fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+    current_memory_bytes: usize,
+    /// The maximum allowable memory usage in bytes.
+    max_memory_bytes: usize,
+    /// Whether to include values in debug logs
+    log_values: bool,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: Ord
+        + Clone
+        + std::fmt::Display
+        + std::hash::Hash
+        + std::str::FromStr
+        + std::fmt::Debug,
+    V: Clone + std::fmt::Debug + AsRef<[u8]> + From<Vec<u8>>,
+{
+    /// Creates a new cache instance.
+    ///
+    /// # Arguments
+    /// * `label` - Identifier used in logs
+    /// * `path` - Directory where cache files are stored
+    /// * `max_memory_bytes` - Maximum bytes to keep in memory
+    /// * `log_values` - Whether to include values in debug logs
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+        log_values: bool,
+    ) -> Result<Self> {
+        let mut cache = Self {
+            label: label.clone(),
+            path: PathBuf::from(path),
+            // TODO: NEED FIX
+            memory_cache: LruCache::new(
+                NonZeroUsize::new(max_memory_bytes)
+                    .expect("Capacity can't be zero"),
+            ),
+            current_memory_bytes: 0,
+            max_memory_bytes,
+            log_values,
+        };
+
+        log::debug!(
+            "cache/{}: initialized with {} bytes limit",
+            label,
+            max_memory_bytes
+        );
+
+        cache.load_fs()?;
+        Ok(cache)
+    }
+
+    /// Retrieves a value by its key, checking memory first then disk.
+    /// Returns None if the key doesn't exist.
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        log::debug!("cache/{}: retrieving value for key {}", self.label, key);
+
+        let value = self
+            .fetch_from_memory(key)
+            .or_else(|| self.fetch_from_disk(key));
+
+        match &value {
+            Some(v) => {
+                if self.log_values {
+                    log::debug!(
+                        "cache/{}: found value for key {}: {:?}",
+                        self.label,
+                        key,
+                        v
+                    );
+                }
+            }
+            None => {
+                log::warn!(
+                    "cache/{}: no value found for key {}",
+                    self.label,
+                    key
+                );
+            }
+        }
+
+        value
+    }
+
+    /// Stores a new value with the given key.
+    /// Returns error if the key already exists or if writing fails.
+    pub fn set(&mut self, key: K, value: V) -> Result<()> {
+        log::debug!("cache/{}: setting value for key {}", self.label, key);
+
+        // Check if value already exists
+        if self.exists(&key) {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                format!("Key {} already exists in cache", key),
+            ));
+        }
```

Comment:
I think it’s an intuitive feature for a `set` operation to overwrite the value if it already exists. It might be better if the developer at least had the option to decide whether to overwrite the value or not.